CIScripts/Common/Invoke-UntilSucceeds.ps1

```diff
@@ -0,0 +1,56 @@
+. $PSScriptRoot\Aliases.ps1
+. $PSScriptRoot\Exceptions.ps1
+
+function Invoke-UntilSucceeds {
+    <#
+    .SYNOPSIS
+    Utility wrapper for making sure that the assert is consistently true. It works by retrying
+    the assert every Interval seconds, up to Duration.
+    .PARAMETER ScriptBlock
+    ScriptBlock containing a Pester assertion.
+    .PARAMETER Interval
+    Interval (in seconds) between retries.
+    .PARAMETER Duration
+    Timeout (in seconds).
+    #>
+    Param (
+        [Parameter(Mandatory=$true,
+                   ValueFromPipeline=$true)] [ScriptBlock] $ScriptBlock,
+        [Parameter(Mandatory=$false)] [int] $Interval = 1,
+        [Parameter(Mandatory=$false)] [int] $Duration = 3
+    )
+    if ($Duration -lt $Interval) {
+        throw "Duration must be longer than interval"
+    }
+    if ($Interval -eq 0) {
+        throw "Interval must not be equal to zero"
+    }
+    $HasPrintedDots = $false
+    $StartTime = Get-Date
+    while (((Get-Date) - $StartTime).Seconds -lt $Duration) {
+        try {
+            $ReturnVal = & $ScriptBlock
+            if ($ReturnVal) {
+                break
```

Comment:
This contract is not obvious - what does it mean for the command to "succeed"? Is returning `$false` a failure, or just successful return of `$false`?

Reply:
What do you mean it's not obvious, there are test cases that cover it pretty extensively?

Reply:
I mean that command could succeed and yet return `$false`, but now we interpret the `$false`/`$null` or `0` unconditionally as a failure (I know, it's documented, but it's not obvious definition of `Succeed`). I'll accept this function as is, but I still think it does two things at once (checking exceptions ∧ checking value), which violates single responsibility principle. I think that the following design would be more elegant:

Case 1, user considers `0` as valid result:
```
$Count = Retry-UntilSucceeds { Count-Things }
```

Case 2, User considers `$null` as an error:
```
$TheThing = Retry-UntilSucceeds {
    Get-Thing | ThrowOnFalse
}
```

(where `ThrowOnFalse` is helper function – an identity function that additionally performs the check)

Reply:
I disagree that it violates SRP. "This function invokes the command until it evaluates to true". It reduces a lot of boilerplate all over the code. It's an "if" on steroids.

Case 1: valid, but this function isn't about counting things but retrying one command.

Case 2: $null is treated as an error, and retries will continue.